File: dart/sdk/lib/src/protocol_settings.dart
import 'package:fixnum/fixnum.dart';
import 'package:rational/rational.dart';

class ProtocolSettings {
  final Rational fEffective;
  final int vrfPrecision;
  final Rational vrfAmplitude;
  final int chainSelectionKLookback;
  final Duration slotDuration;

  ProtocolSettings({
    required this.fEffective,
    required this.vrfPrecision,
    required this.vrfAmplitude,
    required this.chainSelectionKLookback,
    required this.slotDuration,
  });

  factory ProtocolSettings.fromMap(Map<String, String> map) => ProtocolSettings(
        fEffective: _parseRational(map["f-effective"]!),
        vrfPrecision: int.parse(map["vrf-precision"]!),
        vrfAmplitude: _parseRational(map["vrf-amplitude"]!),
        chainSelectionKLookback: int.parse(map["chain-selection-k-lookback"]!),
        slotDuration:
            Duration(milliseconds: int.parse(map["slot-duration-ms"]!)),
      );

  ProtocolSettings mergeFromMap(Map<String, String> map) => ProtocolSettings(
        fEffective: map.containsKey("f-effective")
            ? _parseRational(map["f-effective"]!)
            : fEffective,
        vrfPrecision: map.containsKey("vrf-precision")
            ? int.parse(map["vrf-precision"]!)
            : vrfPrecision,
        vrfAmplitude: map.containsKey("vrf-amplitude")
            ? _parseRational(map["vrf-amplitude"]!)
            : vrfAmplitude,
        chainSelectionKLookback: map.containsKey("chain-selection-k-lookback")
            ? int.parse(map["chain-selection-k-lookback"]!)
            : chainSelectionKLookback,
        slotDuration: map.containsKey("slot-duration-ms")
            ? Duration(milliseconds: int.parse(map["slot-duration-ms"]!))
            : slotDuration,
      );

  static const defaultAsMap = {
    "f-effective": "3/25",
    "vrf-precision": "40",
    "vrf-amplitude": "6/15",
    "chain-selection-k-lookback": "576",
    "slot-duration-ms": "1000",
  };

  @override
  String toString() =>
      "ProtocolSettings(fEffective=$fEffective, vrfPrecision=$vrfPrecision, vrfAmplitude=$vrfAmplitude, kLookback=$chainSelectionKLookback slotDuration=${slotDuration.inMilliseconds}ms, epochLength=$epochLength)";

  static final ProtocolSettings defaultSettings =
      ProtocolSettings.fromMap(defaultAsMap);

  int get chainSelectionSWindow =>
      (Rational(BigInt.from(chainSelectionKLookback), BigInt.from(4)) *
              fEffective.inverse)
          .round()
          .toInt();

  Int64 get epochLength =>
      Int64((Rational(BigInt.from(chainSelectionKLookback)) *
              fEffective.inverse *
              Rational(BigInt.from(3)))
          .round()
          .toInt());
}

_parseRational(String value) {
  final split = value.split("/");
  if (split.length == 1)
    return Rational(BigInt.parse(split[0]));
  else
    return Rational(BigInt.parse(split[0]), BigInt.parse(split[1]));
}

File: scala/node/src/main/scala/com/giraffechain/consensus/ProtocolSettings.scala
package com.giraffechain.consensus

import cats.Show
import cats.implicits.showInterpolator
import com.giraffechain.utility.Ratio

import scala.concurrent.duration.{FiniteDuration, given}

case class ProtocolSettings(
    fEffective: Ratio,
    vrfPrecision: Int,
    vrfAmplitude: Ratio,
    chainSelectionKLookback: Int,
    slotDuration: FiniteDuration
):

  val chainSelectionSWindow: Int =
    (Ratio(chainSelectionKLookback, 4) * fEffective.inverse).round.toInt

  val epochLength: Int =
    (Ratio(chainSelectionKLookback * 3) * fEffective.inverse).round.toInt

  require(epochLength % 3L == 0, s"Epoch length=$epochLength must be divisible by 3")

  override def toString: String =
    show"ProtocolSettings(" +
      show"fEffective=$fEffective," +
      show" vrfPrecision=$vrfPrecision," +
      show" vrfAmplitude=$vrfAmplitude," +
      show" kLookback=$chainSelectionKLookback," +
      show" slotDuration=$slotDuration," +
      show" epochLength=$epochLength slots" +
      show")"

  def merge(map: Map[String, String]): ProtocolSettings =
    map.foldLeft(this)(_.withSetting.apply.tupled(_))

  def withSetting(name: String, value: String): ProtocolSettings =
    name match {
      case "f-effective"                => copy(fEffective = parseRational(value))
      case "vrf-precision"              => copy(vrfPrecision = value.toInt)
      case "vrf-amplitude"              => copy(vrfAmplitude = parseRational(value))
      case "chain-selection-k-lookback" => copy(chainSelectionKLookback = value.toInt)
      case "slot-duration-ms"           => copy(slotDuration = value.toLong.milli)
      case _                            => throw IllegalArgumentException(name)
    }
  private def parseRational(value: String): Ratio =
    value.split('/') match {
      case Array(numerator)              => Ratio(BigInt(numerator))
      case Array(numerator, denominator) => Ratio(BigInt(numerator), BigInt(denominator))
      case _                             => throw IllegalArgumentException(value)
    }
  def toMap: Map[String, String] =
    Map(
      "f-effective" -> fEffective.toString,
      "vrf-precision" -> vrfPrecision.toString,
      "vrf-amplitude" -> vrfAmplitude.toString,
      "chain-selection-k-lookback" -> chainSelectionKLookback.toString,
      "slot-duration-ms" -> slotDuration.toMillis.toString
    )

object ProtocolSettings:
  val Default: ProtocolSettings = ProtocolSettings(
    fEffective = Ratio(1, 50),
    vrfPrecision = 40,
    vrfAmplitude = Ratio(1, 40),
    chainSelectionKLookback = 576,
    slotDuration = 500.milli
  )

  given Show[ProtocolSettings] = Show.fromToString
